connector/src/yang/connector/grpc.py

```diff
@@ -0,0 +1,179 @@
+
+
+# create a logger for this module
+log = logging.getLogger(__name__)
+
+
+class Grpc(BaseConnection):
+    """Session handling for Grpc outbound connections.
+
+        Can be used with pyATS same as yang.connector.Netconf is used or
+        can be used as a standalone module.
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.device = kwargs.get('device')
+        dev_args = self.connection_info
+        self.log = log
+        self.log.setLevel(logging.INFO)
+
+        if protocol := dev_args.get('protocol', '').lower() != 'grpc':
+            msg = f"Invalid protocol {protocol}"
+            raise TypeError(msg)
+
+        self.username = dev_args.get('username', '')
+        self.password = dev_args.get('password', '')
+        if not self.username or not self.password:
+            credentials = dev_args.get('credentials', '')
+            if not credentials:
+                raise KeyError("No credentials found for testbed")
+            if 'grpc' not in credentials:
+                log.info(f"Credentials used from {next(iter(credentials))}")
+            grpc_uname_pwd = credentials.get('')
+            if not grpc_uname_pwd:
+                raise KeyError('No credentials found for gRPC testbed')
+
+        self.host = dev_args.get('host') or dev_args.get('ip')
+
+        self.transporter = dev_args.get('transporter', 'telegraf').lower()
+        self.output_file = dev_args.get('output_file', f'{runtime.directory}/mdt')
+        self.config_file = dev_args.get('config_file', None)
+        self.telemetry_subscription_id = dev_args.get('telemetry_subscription_id', 11172017)
+
+        self.telegraf_process = None
+
+    @property
+    def connected(self):
+        """Return True if session is connected."""
+        return True
+
+    def connect(self):
+        allocated_port = None
+
+        # Allocate a random available port to localhost
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as grpc_socket:
+            grpc_socket.bind(('localhost', 0))
+            _, allocated_port = grpc_socket.getsockname()
+
+            # run config generation within context manager to hold port until it can be passed to telegraf
+            if self.transporter == 'telegraf':
+                config = configparser.ConfigParser()
+                if self.config_file:
+                    # load configuration file
+                    config.read(self.config_file)
+                    if '[inputs.cisco_telemetry_mdt]' not in config.sections():
+                        config.add_section('[inputs.cisco_telemetry_mdt]')
+
+                    # update input socket listener
+                    config.set('[inputs.cisco_telemetry_mdt]', 'transport', '"grpc"')
+                    config.set('[inputs.cisco_telemetry_mdt]', 'service_address', f'":{allocated_port}"')
+
+                    # write configuration file
+                    with open(self.config_file, 'w') as f:
```

Comment:
Please copy the file to runtime.directory (assuming the user provided config is not the same), don't overwrite existing files or change existing user provided file.